File: interpreter-for-cpp/src/VirtualMachine.cpp

#include "VirtualMachine.h"
#include "Base/ErrorLogger.h"
#include "Runtime/Executer.h"
#include "Runtime/ModulePool.h"
#include "Runtime/System.h"
#include "VirtualTool.h"

using namespace peak;
using namespace peak::interpreter;

std::shared_ptr<VirtualJourney> VirtualMachine::Load(const std::string& src) {
	auto executer = Executer::Create(src);
	if (executer) {
		return std::shared_ptr<VirtualJourney>(new VirtualJourney(executer));
	}
	return nullptr;
}

std::shared_ptr<VirtualJourney> VirtualMachine::LoadFile(const std::string& filename) {
	const auto& src = VirtualTool::OpenSrc(filename);
	return Load(src);
}

void VirtualMachine::LocateLogger(std::function<void(const std::string&)> logger) {
	LocateLogger(logger, logger);
}
void VirtualMachine::LocateLogger(std::function<void(const std::string&)> echo, std::function<void(const std::string&)> errorLogger) {
	System::LocateEcho(echo);
	ErrorLogger::Locate(errorLogger);
}

void VirtualMachine::LocateOpenSrc(std::function<std::string(const std::string&)> func) {
	System::LocateOpenSrc(func);
}

ModulePool* VirtualMachine::GetModulePool() {
	return ModulePool::GetInstance();
}

File: interpreter-for-cs/src/VirtualMachine.cs
/*
	PeakScript: https://github.com/Jenocn/PeakScript
	By Jenocn: https://jenocn.github.io
*/

using System;

namespace peak {
	using interpreter;
	public class VirtualMachine {
		public static VirtualJourney Load(string src, Space parent = null) {
			var parseData = ParseTool.Load(src);
			if (!parseData.bSuccess) {
				return null;
			}
			return new VirtualJourney(parseData.sentenceList, parent);
		}
		public static VirtualJourney LoadFile(string filename, Space parent = null) {
			var src = VirtualTool.OpenFile(filename);
			return Load(src, parent);
		}
		public static void LocateLogger(Action<string> logger) {
			LocateLogger(logger, logger);
		}
		public static void LocateLogger(Action<string> echo, Action<string> errorLogger) {
			System0.LocateEcho(echo);
			ErrorLogger.Locate(errorLogger);
		}
	}
} // namespace peak